transports/noise/src/io/framed.rs

```diff
@@ -43,61 +40,94 @@ static_assertions::const_assert! {
     MAX_FRAME_LEN + EXTRA_ENCRYPT_SPACE <= MAX_NOISE_MSG_LEN
 }
 
-/// A `NoiseFramed` is a `Sink` and `Stream` for length-delimited
-/// Noise protocol messages.
-///
-/// `T` is the type of the underlying I/O resource and `S` the
-/// type of the Noise session state.
-pub(crate) struct NoiseFramed<T, S> {
-    io: T,
+/// Codec holds the noise session state `S` and acts as a medium for
+/// encoding and decoding length-delimited session messages.
+pub(crate) struct Codec<S> {
     session: S,
-    read_state: ReadState,
-    write_state: WriteState,
-    read_buffer: Vec<u8>,
-    write_buffer: Vec<u8>,
+    encrypt_buffer: Vec<u8>,
     decrypt_buffer: BytesMut,
 }
 
-impl<T, S> fmt::Debug for NoiseFramed<T, S> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("NoiseFramed")
-            .field("read_state", &self.read_state)
-            .field("write_state", &self.write_state)
-            .finish()
+impl<S: SessionState> Codec<S> {
+    pub(crate) fn new(session: S) -> Self {
+        Codec {
+            session,
+            encrypt_buffer: Vec::new(),
+            decrypt_buffer: BytesMut::new(),
+        }
     }
-}
 
-impl<T> NoiseFramed<T, snow::HandshakeState> {
-    /// Creates a nwe `NoiseFramed` for beginning a Noise protocol handshake.
-    pub(crate) fn new(io: T, state: snow::HandshakeState) -> Self {
-        NoiseFramed {
-            io,
-            session: state,
-            read_state: ReadState::Ready,
-            write_state: WriteState::Ready,
-            read_buffer: Vec::new(),
-            write_buffer: Vec::new(),
-            decrypt_buffer: BytesMut::new(),
+    fn encode_bytes(&mut self, item: &[u8], dst: &mut BytesMut) -> Result<(), io::Error> {
+        self.encrypt_buffer
+            .resize(item.len() + EXTRA_ENCRYPT_SPACE, 0);
+        let n = match self.session.write_message(item, &mut self.encrypt_buffer) {
+            Ok(n) => n,
+            Err(e) => {
+                error!("encryption error: {:?}", e);
+                return Err(io::ErrorKind::InvalidData.into());
+            }
+        };
+
+        self.encrypt_buffer.truncate(n);
+
+        dst.put_u16(n as u16);
+        dst.put(&self.encrypt_buffer[..]);
+
+        Ok(())
+    }
+
+    fn decode_bytes(&mut self, src: &mut BytesMut) -> Result<Option<Bytes>, io::Error> {
+        if src.len() < 2 {
+            return Ok(None);
         }
+
+        let len = u16::from_be_bytes([src[0], src[1]]) as usize;
+        if len == 0 || src.len() < len + 2 {
+            return Ok(None);
+        }
```

Comment:
I just stumbled over https://github.com/mxinden/asynchronous-codec/blob/master/src/codec/length.rs. What do you think of using it instead of re-implementing it here?

Reply:
Sure can. It would make the code much cleaner too!